### ISSUE TITLE
Extension not installed after installing linux packages (firefox)

### ISSUE THREAD

**Issue by maltfield** [web-eid/web-eid-webextension#41]
Issue: It appears that the `Web eID` extension does *not* get installed in Firefox after installing all the relevant packages in Debian 10.

## What is expected

It's expected that, after following the install instructions that install the following packages:

1. open-eid
2. web-eid-native
3. web-eid
4. web-eid-firefox
5. web-eid-chrome

That I can go here, click the "Authenticate" button, and it will authenticate me in the web browser using my Smart Card

* https://web-eid.eu/

## What actually happens

After installing the above packages (and, yes, restarting firefox), I get the following error when clicking "Authenticate" at the above test URL

```
Authentication failed

[Code]
ERR_WEBEID_EXTENSION_UNAVAILABLE

[Message]
Web-eID extension is not available
```

## Steps to reproduce

1. Install Debian 10
2. Download and execute the `install-open-eid.sh` script https://installer.id.ee/media/install-scripts/install-open-eid.sh
3. Confirm that the latest packages are installed `web-eid`
4. Restart firefox
5. In firefox, go to https://web-eid.eu/
6. Click the `Authenticate` button
7. Get error message

**Comment by maltfield** [comment]
I encountered this issue today when attempting to sign a report on RIK's website

 * https://ariregister.rik.ee/eng

After a lot of debugging, I saw that I had all of the packages that I need, and that this *should* be working properly

```
user@estonia:~$ sudo dpkg -l | grep eid
ii  open-eid                              22.6.0.1926-1804                        all          This is eID Software metapackage
ii  web-eid                               2.0.0.552-1804                          all          Web eID application suite (metapackage)
ii  web-eid-chrome                        2.0.2.565-1804                          all          Web eID browser extension for Chrome
ii  web-eid-firefox                       2.0.2.565-1804                          all          Web eID browser extension for Firefox
ii  web-eid-native                        2.0.2.565-1804                          amd64        Web eID browser extension helper application
user@estonia:~$ 
```

I did some more digging into the files installed by these packages

```
user@estonia:~/.mozilla/firefox$ dpkg-query -L web-eid
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/web-eid
/usr/share/doc/web-eid/changelog.Debian.gz
/usr/share/doc/web-eid/copyright
user@estonia:~/.mozilla/firefox$ 

user@estonia:~/.mozilla/firefox$ dpkg-query -L web-eid-native
/.
/usr
/usr/bin
/usr/bin/web-eid
/usr/share
/usr/share/applications
/usr/share/applications/web-eid.desktop
/usr/share/doc
/usr/share/doc/web-eid-native
/usr/share/doc/web-eid-native/changelog.Debian.gz
/usr/share/doc/web-eid-native/copyright
/usr/share/icons
/usr/share/icons/hicolor
/usr/share/icons/hicolor/128x128
/usr/share/icons/hicolor/128x128/apps
/usr/share/icons/hicolor/128x128/apps/web-eid.png
/usr/share/icons/hicolor/16x16
/usr/share/icons/hicolor/16x16/apps
/usr/share/icons/hicolor/16x16/apps/web-eid.png
/usr/share/icons/hicolor/256x256
/usr/share/icons/hicolor/256x256/apps
/usr/share/icons/hicolor/256x256/apps/web-eid.png
/usr/share/icons/hicolor/32x32
/usr/share/icons/hicolor/32x32/apps
/usr/share/icons/hicolor/32x32/apps/web-eid.png
/usr/share/icons/hicolor/512x512
/usr/share/icons/hicolor/512x512/apps
/usr/share/icons/hicolor/512x512/apps/web-eid.png
user@estonia:~/.mozilla/firefox$ 

user@estonia:~/.mozilla/firefox$ dpkg-query -L web-eid-firefox
/.
/usr
/usr/lib
/usr/lib/mozilla
/usr/lib/mozilla/native-messaging-hosts
/usr/lib/mozilla/native-messaging-hosts/eu.webeid.json
/usr/share
/usr/share/doc
/usr/share/doc/web-eid-firefox
/usr/share/doc/web-eid-firefox/changelog.Debian.gz
/usr/share/doc/web-eid-firefox/copyright
user@estonia:~/.mozilla/firefox$ 
```

Of particular note was these

1. /usr/bin/web-eid
2. /usr/lib/mozilla/native-messaging-hosts/eu.webeid.json

The first was just a binary. When I open it, it says 

```
This is a companion application to the Web eID browser extension and cannot be used indepedently. To uninstall the extension, remove this application.
```

The second one is a simple json

```
user@estonia:~/.mozilla/firefox$ cat /usr/lib/mozilla/native-messaging-hosts/eu.webeid.json
{
"name": "eu.webeid",
"description": "Web-eid native application",
"path": "/usr/bin/web-eid",
"type": "stdio",
"allowed_extensions": [ "{e68418bc-f2b0-4459-a9ea-3e72b6751b07}" ]
}
user@estonia:~/.mozilla/firefox$ 
```

I did some research and found that a Native Messaging Host is basically an binary that lives on your computer that speaks to a website through the browser using json messages. So the above file tells firefox (mozilla actually) about this "Native Messaging Host" app named "eu.webeid"

 * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging

I couldn't find a way to enumerate all of the "Native Messaging Hosts" recognized by my browser, so I created this question on SuperUser:

 * https://superuser.com/questions/1732697/how-do-i-list-absolutely-all-of-the-extensions-addons-plugins-etc-currently-i

In any case, it appears that the `/usr/lib/mozilla/native-messaging-hosts/` directory was correct for a linux system per the documentation:

 * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#linux

I noticed that the json file itself was pretty bare. It specified the path to the binary `/usr/bin/web-eid` and that the (json) communication should happen with `stdio`. I know I have that binary and it's working (see above), but what about this extension `"{e68418bc-f2b0-4459-a9ea-3e72b6751b07}"`?

As mentioned in another ticket for another plugin related to signing things with my Estonian ID card, I only saw one extension in `about:support` and `about:addons` = `PKCS 11 loader`

 * https://github.com/open-eid/firefox-pkcs11-loader/issues/36

I held my nose and searched for firefox extensions named `open-eid`, and I found this

 * https://addons.mozilla.org/en-US/firefox/addon/web-eid-webextension/

After installing it, I confirmed that the uuid of the extension matches the one in the whitelist of the json above

```
user@estonia:~$ ls ~/.mozilla/firefox/*/extensions
{e68418bc-f2b0-4459-a9ea-3e72b6751b07}.xpi
user@estonia:~$ 
```

I tried the test again, and it worked! I got a window asking me for my PIN instead of an error message.

Anyway, this is really insecure. Installing packages through apt has the added security of the repo's cryptographically signed manifest file that ensures the authenticity and integrity of all packages installed in `apt` -- a much, much higher level of security to installing a package from an online https site with auth limited only to the X.509 PKI.

Please fix this so that installing the `web-eid-firefox` package in `apt` actually installs the extension in the firefox browser.

**Comment by metsma** [comment]
> Anyway, this is really insecure. Installing packages through apt has the added security of the repo's cryptographically signed manifest file that ensures the authenticity and integrity of all packages installed in apt -- a much, much higher level of security to installing a package from an online https site with auth limited only to the X.509 PKI.
This is something we cannot control. Firefox and Chrome enforce to install all extensions from add-on stores.
Least Chrome has option to give hint install extension automatically from store.
With Firefox there is only option use enterprise policy to install extension from store (seems like this is current issue)

**Comment by maltfield** [comment]
Sorry for my skepticism, but how do these packages install firefox extensions then?

```
user@disp9032:~$ sudo apt-cache search firefox | grep -iE 'extension|addon'
webext-browserpass - web extension for the password manager pass
chrome-gnome-shell - GNOME Shell extensions integration for web browsers
webext-form-history-control - extension to manage form history
webext-https-everywhere - Extension to force the use of HTTPS on many sites
webext-keepassxc-browser - Web browser extension to organize web site credentials in KeePassXC
libhtml-widgets-selectlayers-perl - Perl extension for selectable HTML layers
mozilla-devscripts - Development scripts used by Mozilla's addons packages
user@disp9032:~$ 
```

**Comment by metsma** [comment]
> https://packages.debian.org/bullseye/all/chrome-gnome-shell/filelist   
It does not contain extension.
> webext-*  
Probably debian patched firefox to allow sideloading extensions
https://blog.mozilla.org/addons/2020/03/10/support-for-extension-sideloading-has-ended/

**Comment by maltfield** [comment]
ok, that's unfortunate. Is there any UX solution to this then?

It took me (a software engineer with a GitHub account and the support of a couple developers [thank you!] and reading the architectural documentation) a whole day to figure out I couldn't sign my report on an RIA website because I was missing the web browser extension. At no point did any error message pop-up telling me to install the extension (and telling me that installing the `web-eid-firefox` package was not enough, and I would additionally need to install the extension manually).

There's a lot of moving pieces here, and if one (or many) of the components are missing, I believe the UI should be very clear to the user exactly what is missing and what they need to do to fix it.

**Comment by metsma** [comment]
> Does that mean it works for Ubuntu as well, which is a derivative of Debian and supported by this project? Can this web-eid extension be sideloaded by the web-eid-firefox Ubuntu package as a resolution to this ticket?

Not really. Ubuntu 22.04 is moving away deb packaged Firefox (to snap) and this has other set problems (Sandbox does not allow to communicate with native messaging host and there is currently no workaround, except to install official Firefox/Chrome binary)

> And is there any UX solution to this?

There is no easy solution. Only UI user has at this situation is web service html. HTML site can only verify if there is extension installed or not (read browser restrictions). If extension is installed we can use javascript console for debug messages.
Not sure how to log or debug extension installation failures. Probably using Firefox debug logs or similar, where can probably see wrong paths, http/https download issues etc.

**Comment by maltfield** [comment]
> Ubuntu 22.04 is moving away deb packaged Firefox (to snap) and this has other set problems (Sandbox does not allow to communicate with native messaging host and there is currently no workaround, except to install official Firefox/Chrome binary)

Oh man, you're right. When I saw that firefox switched from deb to snap I decided to stop recommending Ubuntu and recommend Mint instead.

snap is a cancer closed-wall garden created by Canonical, and a lot of people are leaving Ubuntu because of this. I highly, highly recommend supporting Debian instead of Ubuntu; you'll support more users, especially as everyone leaves Ubuntu for more libre distros.

Is "supporting Ubuntu only" RIK's requirement? Can it be changed through an enumeration of:

1. technical documentation of all of the issues with Ubuntu and snaps ([Mint has some strong opinions about snaps](https://linuxmint-user-guide.readthedocs.io/en/latest/snap.html))
2. documentation of the[ market share of Ubuntu declining vs Debian and Debian-derived systems (including Mint)](http://www.geeks3d.com/20111124/linux-distributions-mint-ahead-of-ubuntu/) surpassing it
3. etc?
